### ISSUE TITLE
Reintroduce support for Python 3.7

### ISSUE THREAD

**Issue by sobolevn** [antonagestam/phantom-types#122]
Hi! A quick question:

Why have you dropped `python3.7` support in this commit https://github.com/antonagestam/phantom-types/commit/c0817edd886237d288b277c93a89c04d286e54e1 ?

Looks like `python3.7` was supported not a long time ago: https://github.com/antonagestam/phantom-types/commit/c2e6d37a0afa2a9dd531df016414a12688341006

From what I can see the only feature of `python3.8` you are using is `/` arguments, which can be easily converted into regular arguments (which is even cleaner in my opinion). Anything else that I've missed?

**Comment by antonagestam** [comment]
@sobolevn I think it was related to ProtocolMeta, but I can't remember the specific details, it might be the `/` that you mentioned too.

More generally however, I felt that I wanted to prioritize being able to iterate fast and being able to focus on working out ergonomic APIs in the early stages. Perhaps naïvely, but I'm thinking the intersection between projects that has the ability to try out something like phantom-types and projects that aren't able to upgrade to Python 3.8+ isn't that common. Of course, that reasoning doesn't take libraries that want to use phantom-types as a dependency and needs to keep support for 3.7 into account. Is that your situation? Basically I think the choice was made mostly because I wanted to spend as much time as possible on exploring what Phantom types in Python could look like and didn't want to think about backwards compatibility.

> From what I can see the only feature of python3.8 you are using is / arguments, which can be easily converted into regular arguments (which is even cleaner in my opinion).

Well, those aren't added for cleanliness, but for explicitness. I'm not saying this is a big issue but if Python 3.7 support is added it would then require a future major release to add these back as limiting to positional only is a breaking change. I do have a quite strong opinion that the code _is_ better with them than without them. Also note that, having `/` in place means that the implementations are free to rename the value parameter and being guaranteed not to break call sites.

As you probably can tell, I wouldn't be very excited to bring back support for Python 3.7, but I'm not necessarily completely against it either.

**Comment by antonagestam** [comment]
@sobolevn If you think adding Python 3.7 compatibility means more usage, it's definitely on the table for discussion! 🙂

**Comment by sobolevn** [comment]
> If you think adding Python 3.7 compatibility means more usage, it's definitely on the table for discussion! 

Yes, I think that `python3.7` is widely popular!
Here are some data: I am checking `requests` stats on https://pypistats.org/packages/requests 
<img width="951" alt="Снимок экрана 2021-07-06 в 15 15 56" src="https://user-images.githubusercontent.com/4660275/124598359-2ea07000-de6d-11eb-9ab6-b6d875a4ca21.png">

From here we can see that `python3.7` is the most used Python right now.

I also don't struggle with `3.7` at all, from my experience it does not add any extra maintaince work. (well, there are some cases like `ast.py` changes or `asyncio`, etc, but it is not related to this project)

>  Of course, that reasoning doesn't take libraries that want to use phantom-types as a dependency and needs to keep support for 3.7 into account. Is that your situation?

Yes, this is exactly the case for me. I am trying to add `phantom-types` as a test dependency, but I cannot do that:

```
» poetry add --dev phantom-types
Using version ^0.9.1 for phantom-types

Updating dependencies
Resolving dependencies... (0.4s)
Resolving dependencies... (0.5s)
  SolverProblemError

  The current project's Python requirement (>=3.7,<4.0) is not compatible with some of the required packages Python requirement:
    - phantom-types requires Python >=3.8, so it will not be satisfied for Python >=3.7,<3.8
  
  Because no versions of phantom-types match >0.9.1,<0.10.0
   and phantom-types (0.9.1) requires Python >=3.8, phantom-types is forbidden.
  So, because classes depends on phantom-types (^0.9.1), version solving failed.

  at ~/.poetry/lib/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│ 
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties
    
    For phantom-types, a possible solution would be to set the `python` property to ">=3.8,<4.0"

    https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies,
    https://python-poetry.org/docs/dependency-specification/#using-environment-markers

```

I can get around this problem via some hacks here and there, but I don't want to, because I would have to put a large "warning" sign in docs ([link](https://github.com/dry-python/classes/pull/256/files#diff-8e639d41f20036807add4142f0cfef3bf25745813357e100c745bd151a3773c2R146)) as well saying "only for python3.8+". 

So, I would love to contribute `python3.7` support, because it would help me a lot and I hope won't cause much (*any*) problems.

**Comment by antonagestam** [comment]
@sobolevn I'm not convinced those numbers are relevant, but I think your use-case makes sense, let's go for it! 🙂

**Comment by orsinium** [comment]
TL;DR: I agree with OP, Python 3.7 is widely used and nice to support.

The long version: You underestimate the inertness of the enterprise 🙃 We're still on Python 3.6 and going to migrate on Python 3.7 soon, just because 3.6 is reaching EoL this December. Our motivation is that we deliver a medical device (hence the interest in phantom types and other means to improve reliability) that heavily relies on ML. So, every change in the product requires a long careful validation of correctness and performance, and changing the Python version can't be done without putting in a lot of effort.

Thank you for the project. I'm wondering if I can generate phantom types in [deal](https://github.com/life4/deal) under the hood from provided by the user contracts.

**Comment by antonagestam** [comment]
@orsinium A bunch of discussion here, nice! 😄

However, I wasn't questioning whether or not Python 3.7 is popular, I know it is!

What I'm questioning is whether the intersection of projects that can bring in an experimental tool like this library (the only production usage I know of is my own, and it's still quite limited), and the set of projects that can't upgrade Python, is a large set.

In my experience, upgrading from Python 3.7 to 3.8 and 3.9 has been extremely smooth with no problems whatsoever for me personally. Introducing phantom-types into a project is a huge change in comparison. So my reasoning is that most projects that have the ability to bring on phantom-types probably have the ability to upgrade Python. I appreciate your feedback and take it as a datapoint that the project you are working on actually is in that intersection 🙂

> So, every change in the product requires a long careful validation of correctness and performance, and changing the Python version can't be done without putting in a lot of effort.

Why don't you upgrade to the most recent version while you're at it? 🙂 The same amount of work will give you 2 more years of support.

> I'm wondering if I can generate phantom types in deal under the hood from provided by the user contracts.

It's my opinion that basically every implementation of "design by contract" in Python goes about it in insufficient ways as they don't interact with the type system. Taking the type system into account, everything becomes a question about "how do I express this as type?", or "how do I narrow this type?". Using a contracts library or even just assertions works for runtime, but there's no way for Python type checkers to work with that. As soon as you've found a way to narrow the type of function arguments properly, verifying those contracts at runtime is a job best left to tools like [beartype](https://pypi.org/project/beartype/) and [typeguard](https://pypi.org/project/typeguard/). A setup like that will give you identical runtime guarantees to using a contracts library, but will also allow static type checking.

**Comment by antonagestam** [comment]
@orsinium Taking the example from the deal readme, the problem is that the post condition isn't expressed as a type. Rewriting it with phantom-types and typeguard, we can do something like this:

```python
from phantom.numeric import Natural
from typeguard import typechecked
from typing import cast

@typechecked
def count(items: List[str], item: str) -> Natural:
    return cast(Natural, items.count(item))
```

But if we always want to assert the post condition, we might as well skip typeguard and just parse into `Natural`:


```python
from phantom.numeric import Natural

def count(items: List[str], item: str) -> Natural:
    return Natural.parse(items.count(item))
```

**Comment by orsinium** [comment]
> Why don't you upgrade to the most recent version while you're at it?

Sure, we will try our best to do the full leap, hard to say what we will face there.

> It's my opinion that basically every implementation of "design by contract" in Python goes about it in insufficient ways as they don't interact with the type system

Deal is different, we go beyond runtime. We [generate tests](https://deal.readthedocs.io/basic/tests.html), have [static checker for simple cases](https://deal.readthedocs.io/basic/linter.html), and soon will have our own [formal verifier](https://github.com/orsinium-labs/deal-solver). An integration with mypy is in question, though, it provides limited capabilities for what plugins can do.

I love liquid/phantom/refined/dependent types, and contracts are basically about the same more strict formal verification. The only difference is that contracts are easier for the user for the cost of much more complex type checker implementation. Like, "hey, compiler, there is my contract, IDK if it is bounds, subclass, or whatever, just figure it out".

**Comment by antonagestam** [comment]
@orsinium If I've understood what others have said about Eiffel correctly, which seems to be the holy grail of DbC, it actually infers type narrowing from pre- and post conditions. It sounds like that's what you want to do?

To me though, the real beauty of having phantom types is to use them very tightly together with a static type checker, because without any runtime burden the static type checker can tell me which of these are invalid:

```python
def do_with_count(count: Natural):
    ...

do_with_count(1)  # incorrect because not *proven* to be Natural
do_with_count(count(...))  # correct because count() returns Natural
```

> I love liquid/phantom/refined/dependent types, and contracts are basically about the same more strict formal verification.

Phrased differently: phantom types provide a way to express contracts in a language understood by static type checkers, as well as verifiable by runtime type checkers.

> contracts are easier for the user 

I sympathize with that, but trying to achieve the level of integration with the type system that a language like Eiffel has in Python seems like swimming upstream. I'd be very impressed by something like that though. With regards to program structure, the idea of reusing contracts is very compelling though and seams to be lost with a DbC approach. E.g., why do I need to specify over and over again what a `Natural` is?

> I'm wondering if I can generate phantom types in deal under the hood from provided by the user contracts.

I don't understand why you would want to do this if not to use the generated types for static type checking?

**Comment by orsinium** [comment]
In deal, `do_with_count(1)` is correct but `do_with_count(-1)` is not. This is the difference.

```
❯ cat tmp.py                
import deal

def natural(x):
    return x > 0

@deal.pre(natural)
def do_with_count(count: int):
    ...

@deal.pure
def example():
    do_with_count(1)
    do_with_count(-1)

❯ python3.9 -m flake8 tmp.py
tmp.py:13:1: DEAL043 missed marker (io)
tmp.py:16:5: DEAL011 pre contract error (-1)
```

Anyway, I see your point, subclassing is also fun and helpful. The truth is always somewhere in the middle. I like type checking not less than contracts. And it's up to you if you want to use any of it.
